src/mixins/object_stacking.mixin.js

```diff
@@ -76,5 +76,48 @@ fabric.util.object.extend(fabric.Object.prototype, /** @lends fabric.Object.prot
       this.canvas.moveTo(this, index);
     }
     return this;
+  },
+
+  /**
+   *
+   * @param {fabric.Object} other object to compare against
+   * @returns {boolean | undefined} if objects do not share a common ancestor or they are strictly equal it is impossible to determine which is in front of the other; in such cases the function returns `undefined`
+   */
+  isInFrontOf: function (other) {
+    if (this === other) {
+      return undefined;
+    }
+    var ancestors = this.getAncestors().reverse().concat(this);
+    var otherAncestors = other.getAncestors().reverse().concat(other);
+    var i, j, found = false;
+    //  find the common ancestor
+    for (i = 0; i < ancestors.length; i++) {
+      for (j = 0; j < otherAncestors.length; j++) {
+        if (ancestors[i] === otherAncestors[j]) {
+          found = true;
+          break;
+        }
+      }
+      if (found) {
+        break;
+      }
+    }
+    if (!found) {
+      return undefined;
+    }
+    //  compare trees from the common ancestor down
+    var tree = ancestors.slice(i),
+        otherTree = otherAncestors.slice(j),
+        a, b, parent;
+    for (i = 1; i < Math.min(tree.length, otherTree.length); i++) {
+      a = tree[i];
+      b = otherTree[i];
+      if (a !== b) {
+        parent = tree[i - 1];
+        return parent._objects.indexOf(a) > parent._objects.indexOf(b);
+      }
+    }
+    //  happens if a is ancestor of b or vice versa
+    return tree.length > otherTree.length;
```

Comment:
seems to me that the first part of this function is `findCommonAncestors`.
Once a common ancestor is found, to me it seem that all that matter is position index of the next acenstor after teh common one. Who is front of who at this level will also dictacte who is i front later.
In fewer words i don't see the reason for the second loop.
I'll paste here in the comment a tenative replacement tomorrow morning ( now too tired )

Reply:
Added the `strict` option and tests
678f528f16d515cc353c6a6014e207da8f1f9769
3b8b393f740a2bfb81c9eb0737c1a8a335c76118

Reply:
The second loop walks down the common ancestors from the top most one until it finds a fork in the tree and then checks the index of each ancestor at the fork and returns.

Reply:
i still don't get it.
findCommonAncestors seems to return the closest common ancestor.
If this function was:

```js
  isInFrontOf: function (other) {
    if (this === other) {
      return undefined;
    }
    var data = this.findCommonAncestors(other)
    if (!data) {
      return undefined;
    }
    parent = data.ancestors[0];
    // this contains other, is not in front of it
    if (parent === this) return false;
    // other contains this, so this is drawn on top of it
    if (parent === other) return true;
    return parent._objects.indexOf(this) > parent._objects.indexOf(other);
  }
```

What i don't get is why findCommonAncestors and this function need to work in a similar logic but reversed.
findCommonAncestors doesn't reverse the ancestors order and so doesn't need to walk the three for a fork, the common ancestor returned, the [0] of it, is the closest container of both objects.
There may be a hole in my logic, but i don't see it.

Reply:
Think of an asymmetric case.
- canvas
  - a
  - b
     - c
        - d

d.findCommonAncestors(a)

Reply:
I will try to explain the riddle by countering your code.
```
d !== a1
data = {index: 2 ,otherIndex: 1, ancestors:[canvas]}
parent !== this
parent !== other
-1 > -1
```

Reply:
updated